### PR TITLE
Extended crash filtering

### PR DESCRIPTION
**Pull request by arifBurakDemiray**

### CODE REVIEW

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
config option should be under "crashes" sub option.
depr call should name the replacement
no changelog entry

The crash filter should
return boolean and be able to change all fields

File: sdk/src/main/java/ly/count/android/sdk/CrashData.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package ly.count.android.sdk;
+
+import java.util.Map;
+
+/**
+ * This class is used to store crash data.
+ */
+public class CrashData {
+
+    private String stackTrace;
```
Comment (ArtursKadikis) [inline sdk/src/main/java/ly/count/android/sdk/CrashData.java:10]: need "fatal" flag

File: sdk/src/main/java/ly/count/android/sdk/ModuleCrash.java [hunk 2]
```diff
@@ -142,6 +142,43 @@ public void sendCrashReportToQueue(String error, boolean nonfatal, boolean isNat
...
+        Utils.truncateSegmentationValues(combinedSegmentationValues, _cly.config_.maxSegmentationValues, "[ModuleCrash] sendCrashReportToQueueWFilterCallback", L);
+
+        //limit the size of the crash report to 20k characters
+        if (!isNativeCrash) {
+            error = error.substring(0, Math.min(20_000, error.length()));
+        }
+
+        CrashData crashData = new CrashData(error, combinedSegmentationValues, DeviceInfo.getLogs());
+
+        if(globalCrashFilterCallback != null) {
```
Comment (ArtursKadikis) [inline sdk/src/main/java/ly/count/android/sdk/ModuleCrash.java:169]: move user filtering before SDK filtering

File: sdk/src/main/java/ly/count/android/sdk/CrashData.java (hunk 1 above)
Comment (ArtursKadikis) [inline sdk/src/main/java/ly/count/android/sdk/CrashData.java:10]: add crash metrics

File: sdk/src/main/java/ly/count/android/sdk/ModuleCrash.java [hunk 3]
```diff
@@ -142,6 +142,43 @@ public void sendCrashReportToQueue(String error, boolean nonfatal, boolean isNat
...
+            if (globalCrashFilterCallback.filterCrash(crashData)) {
+                L.d("[ModuleCrash] Crash filter found a match, exception will be ignored, [" + error.substring(0, Math.min(error.length(), 60)) + "]");
+                return;
+            }
+        }
```
Comment (ArtursKadikis) [inline sdk/src/main/java/ly/count/android/sdk/ModuleCrash.java:174]: if the crash data was modiffied, then in crash metrics we need to set a custom flag

File: sdk/src/main/java/ly/count/android/sdk/CrashData.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    private Map<String,Object> crashSegmentation;
+    private String breadcrumbs;
+    protected boolean crashChanged = false;
```
Comment (ArtursKadikis) [inline sdk/src/main/java/ly/count/android/sdk/CrashData.java:13]: this needs to be per metric

File: sdk/src/main/java/ly/count/android/sdk/CrashData.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    /**
+     * Set the breadcrumbs of the crash.
+     * @param breadcrumbs the breadcrumbs of the crash.
+     */
+    public void setBreadcrumbs(String breadcrumbs) {
+        this.breadcrumbs = breadcrumbs;
+        crashChanged = true;
+    }
+}
```
Comment (ArtursKadikis) [inline sdk/src/main/java/ly/count/android/sdk/CrashData.java:72]: this needs a protected "getChangedBits"

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Add tests for "CrashData" for its changes and bit generation.

rework new config options to better ones

File: sdk/src/main/java/ly/count/android/sdk/ConfigCrashes.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        return this;
+    }
+
+    /**
+     * For use during testing
+     *
+     * @param checkForDumps whether to check for native crash dumps
+     * @return Returns the same config object for convenient linking
+     */
+    protected synchronized ConfigCrashes checkForNativeCrashDumps(boolean checkForDumps) {
```
[the remark this answers is not in the document]
Comment (arifBurakDemiray) [reply sdk/src/main/java/ly/count/android/sdk/ConfigCrashes.java:32]: it is true by default. So enable logic will not apply here. If I modify it, I must also modify a test

[the remark this answers is not in the document]
Comment (arifBurakDemiray) [reply sdk/src/main/java/ly/count/android/sdk/ConfigCrashes.java]: will delete its editibility

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Merge conflict.
Changelog needs to be updated

File: sdk/src/main/java/ly/count/android/sdk/CrashData.java [hunk 7]
```diff
@@ -166,6 +165,17 @@ protected boolean[] getChangedFields() {
         return changedFields;
     }
 
+    protected String getChangedFieldsAsString() {
```
Comment (ArtursKadikis) [inline sdk/src/main/java/ly/count/android/sdk/CrashData.java:168]: getChangedBitsAsInt

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/ly/count/android/sdk/CrashData.java [hunk 8]
```diff
@@ -12,11 +12,11 @@ public class CrashData {
     private @NonNull String stackTrace;
     private @NonNull Map<String, Object> crashSegmentation;
     private @NonNull List<String> breadcrumbs;
-    private boolean fatal;
+    private boolean handled;
     private @NonNull JSONObject crashMetrics;
     private final String[] checksums = new String[5];
```
Comment (ArtursKadikis) [inline sdk/src/main/java/ly/count/android/sdk/CrashData.java:17]: let us leave a comment, what value should be in each index

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
let us keep fatal

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Add tests

**Pull request merged**